### PR TITLE
Constexpr Point [WIP]

### PR DESCRIPTION
**Pull request by rezarastak**
Another step toward #8003 

In order to make the constructors `constexpr`, I replaced the `Assert(dim == 2)` with `static_assert(dim == 2)`. Because of this change, a number of functions have to be reorganized. In particular, anywhere that uses a `switch` statement over `dim` in the following way has to be slightly rewritten. Unfortunately we cannot use `constexpr if` in a portable way here (can we?).
```c++
switch (dim)
  {
    case 1:
      return Point<dim>(x);
    case 2:
      return Point<dim>(x, y);
    case 3;
      return Point<dim>(x, y, z);
  }
```
Sometimes we have to specialize the templated function for each dim. Other times it is better to define useful functions or methods, such as `extend()` and `project()` that I implemented.

This work ended up being bigger than I expected, but I think we can finish it given some time. I look forward to your feedback. This is WIP, so I will add tests and changelogs.

### CODE REVIEW

File: include/deal.II/base/point.h [hunk 1]
```diff
@@ -501,6 +514,33 @@ Point<dim, Number>::operator=(const Tensor<1, dim, OtherNumber> &p)
...
+}
+
+
+
+template <int dim, typename Number>
+inline DEAL_II_CONSTEXPR DEAL_II_CUDA_HOST_DEV Point<dim - 1, Number>
+                                               Point<dim, Number>::project(const unsigned int coord_axis) const
+{
+  Point<dim - 1, Number> result;
+  for (unsigned int i = 0; i < coord_axis - 1; ++i)
```
Comment (masterleinad) [inline include/deal.II/base/point.h]: This won't work if `coord_axis==0` since `unsigned int (-1)` is much larger than any reasonable `dim` here. Same below.

Comment (rezarastak) [reply include/deal.II/base/point.h]: Thank you for noticing it. I fixed it.

File: include/deal.II/base/point.h [hunk 2]
```diff
@@ -157,8 +154,23 @@ class Point : public Tensor<1, dim, Number>
...
+   * points with <tt>dim!=4</tt> as it would leave some components of the
+   * point coordinates uninitialized (if dim>4) or would not use some
+   * arguments (if dim<4).
+   *
+   * @note This function can also be used in CUDA device code.
+   */
+  DEAL_II_CUDA_HOST_DEV constexpr Point(const Number x,
+                                        const Number y,
+                                        const Number z,
+                                        const Number w);
```
Comment (Rombur) [inline include/deal.II/base/point.h]: Why did you this function?

**Comment by masterleinad** [comment]
> Unfortunately we cannot use `constexpr if` in a portable way here (can we?).

We are using `if constexpr` in some places:
```
include/deal.II/base/utilities.h:1196:    if constexpr (std::is_trivially_copyable<T>() && sizeof(T) < 256)
include/deal.II/base/utilities.h:1278:    if constexpr (std::is_trivially_copyable<T>() && sizeof(T) < 256)
include/deal.II/base/partitioner.templates.h:511:          if constexpr (std::is_trivial<Number>::value)
include/deal.II/base/partitioner.templates.h:721:              if constexpr (std::is_trivial<Number>::value)
include/deal.II/lac/vector_operations_internal.h:237:            if constexpr (std::is_trivial<Number>::value)
include/deal.II/lac/vector_operations_internal.h:274:        if constexpr (std::is_trivially_copyable<Number>() &&
include/deal.II/lac/sparse_matrix.templates.h:224:      if constexpr (std::is_trivial<number>::value)
```
guarding it with `DEAL_II_WITH_CXX17`. 

Given the number of places you need to change, there might be a good chance that this will break a lot of user codes. Maybe it's a good idea to use `DEAL_II_CONSTEXPR` instead of `constexpr` for the constructors if that works better (i didn't check yet).

File: include/deal.II/base/point.h (hunk 2 above)
Comment (rezarastak) [reply include/deal.II/base/point.h]: In order to fix compilation of `mapping_cartesian.cc`, I added the array `GeometryInfo<dim>::unit_normal_vector` as it made the code more readable. Then I noticed that `GeometryInfo<dim>` is also defined for `dim == 4`. So then I had to add a constexpr constructor for `Point<4>`. 

This was my decision process. Let me know your suggestions.

**Comment by rezarastak** [comment]
@masterleinad Using `if constexpr` requires using `#ifdef` and `#else` and repeating the code. It only works well in small code sections. I didn't find it useful in the lines of codes I modified.

For many methods we can use `DEAL_II_CONSTEXPR` but I try my best to create `constexpr` constructors whenever possible. Because of using `constexpr` constructors, I can now define `std::array<Point<dim>, faces_per_cell> GeometryInfo<dim>::unit_normal_vector` without using any macros or code duplications.

My proposed replacement for to following code:
```c++
template<int dim>
std::vector<Point<dim>> get_vector()
{
  std::vector<Point<dim>> result;
  for (int i = 0; i < 5; ++i)
  {
    switch (dim)
    {
      case 1:
        result.push_back(Point<dim>(i));
        break;
      case 2:
        result.push_back(Point<dim>(i, 2 * i));
        break;
      case 3;
        result.push_back(Point<dim>(i, 2 * i, 3 * i ));
    }
  }
  return result;
}
```
is either:
```c++
template <int dim>
Point<dim> get_point_from_index(int i);

template <>
Point<1> get_point_from_index(int i)
{
  return {i};
}

template <>
Point<2> get_point_from_index(int i)
{
  return {i, 2 * i};
}

template <>
Point<3> get_point_from_index(int i)
{
  return {i, 2 * i, 3 * i};
}

template<int dim>
std::vector<Point<dim>> get_vector()
{
  std::vector<Point<dim>> result;
  for (int i = 0; i < 5; ++i)
    result.push_back(get_point_from_index<dim>(i));
  return result;
}
```
or:
```c++
template<int dim>
std::vector<Point<dim>> get_vector()
{
  std::vector<Point<dim>> result;
  for (int i = 0; i < 5; ++i)
  {
    const auto point_all_dim = std::make_tuple(Point<1>(i), Point<2>(i, 2 * i), Point<3>(i, 2 * i, 3 * i));
    result.push_back(std::get<dim - 1>(point_all_dim));
  }
  return result;
}
```
The first option is better for longer dimension dependent code sections, but the second option is easier to use for one-line dimension dependent points. Both options are completely type-safe and they don't use debug mode asserts. If we are willing to modify a small number of `switch` statements (which I have almost finished it), we can go forward with the `constexpr Point<dim>` constructors. 

I used the combination of `std::make_pair` and `std::get` in the tutorials.

File: source/grid/grid_out.cc [hunk 3]
```diff
@@ -4228,6 +4227,7 @@ namespace internal
       {}
     };
 
+    using LineList = std::list<LineEntry>;
```
Comment (bangerth) [inline source/grid/grid_out.cc]: Our usual convention is to avoid these sort of convenience typedefs since they make it more difficult to read code.

File: source/grid/grid_out.cc [hunk 4]
```diff
@@ -4270,6 +4270,169 @@ namespace internal
     }
 
 
+    template <int dim, int spacedim>
+    LineList
+    get_lines(const dealii::Triangulation<dim, spacedim> &,
```
Comment (bangerth) [inline source/grid/grid_out.cc]: put this into a separate PR as well

File: source/numerics/vector_tools_project_qpmf.inst.in [hunk 5]
```diff
@@ -46,4 +46,4 @@ for (VEC : REAL_NONBLOCK_VECTORS; deal_II_dimension : DIMENSIONS)
 
 
     \}
-  }
\ No newline at end of file
+  }
```
Comment (bangerth) [inline source/numerics/vector_tools_project_qpmf.inst.in]: separate PR

File: tests/base/point_04.cc [hunk 6]
```diff
@@ -14,7 +14,7 @@
 // ---------------------------------------------------------------------
 
 
-// test for Point::operator()
+// test for Point::operator() and the constructor from boost::geometry
```
Comment (bangerth) [inline tests/base/point_04.cc]: separate PR

File: examples/step-22/step-22.cc [hunk 7]
```diff
@@ -996,21 +996,24 @@ namespace Step22
...
       std::vector<unsigned int> subdivisions(dim, 1);
       subdivisions[0] = 4;
 
-      const Point<dim> bottom_left = (dim == 2 ?                //
-                                        Point<dim>(-2, -1) :    // 2d case
-                                        Point<dim>(-2, 0, -1)); // 3d case
+      const auto bottom_left_2d_or_3d =
+        std::make_pair(Point<2>(-2, -1),     // 2d case
+                       Point<3>(-2, 0, -1)); // 3d case
+      const Point<dim> bottom_left = std::get<dim - 2>(bottom_left_2d_or_3d);
```
[the remark this answers is not in the document]
Comment (rezarastak) [reply examples/step-22/step-22.cc]: Since I changed the `Assert` statements in `Point<dim>` constructors into `static_assert`, the original statement
```c++
Point<dim> bottom_left = (dim == 2 ? Point<dim>(-2, -1) :  Point<dim>(-2, 0, -1));
```
fails to compile for both `dim == 2` and `dim == 3`. The neatest possible way I found around is using `std::make_pair` or `std::make_tuple`. Do you think this is an acceptable and readable modification?

File: include/deal.II/base/point.h [hunk 8]
```diff
@@ -176,33 +188,47 @@ class Point : public Tensor<1, dim, Number>
...
-  operator=(const Tensor<1, dim, OtherNumber> &p);
+  DEAL_II_CONSTEXPR Point<dim, Number> &
+                    operator=(const Tensor<1, dim, OtherNumber> &p);
+
+  /**
+   * Converts the Point<dim> to a Point<dim + 1> object by appending a
+   * coordinate value as the last coordinate
+   */
+  DEAL_II_CONSTEXPR DEAL_II_CUDA_HOST_DEV Point<dim + 1, Number>
+                                          extend(const Number extension) const;
```
[the remark this answers is not in the document]
Comment (rezarastak) [reply include/deal.II/base/point.h]: Adding this function allows the simplification I made in the file [data_out_stack.cc](https://github.com/dealii/dealii/pull/8673/files#diff-943de848b956aaea862c5d1f5a09bb00)

Within `data_out_stack.cc` we need a non-zero value for the argument `extension`.

The changes and the introduction of the `extend` method are necessary to prevent compile error in `data_out_stack.cc` due to `static_assert` in the constructors of `Point`. Basically, whenever there is a `switch (dim)` statement and it constructs `Point<dim>` within that block, we need to modify it to allow `static_assert` in constructors.

File: examples/step-22/step-22.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply examples/step-22/step-22.cc]: Oh, I see. No, this isn't acceptable -- we use this idiom all over the place, not just in the library or the tutorial, but many user programs as well. All of these would now stop to compile. So that can't work.

You'll have to find a different way to dealing with this than the `static_assert` in the constructor. Sorry :-(

File: include/deal.II/base/point.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/point.h]: I see.

As I mentioned before, I think the addition of the `extend`/`embed` function by itself, and its use in `data_out_stack.cc` makes sense, and could be a patch of itself.

**Comment by rezarastak** [comment]
@bangerth Your suggestion makes perfect sense, I think I should split this into smaller PRs.

My goals in this PR is two-fold:
1. Make the constructors of `Point<dim>` constexpr. There are many benefits for making them `constexpr`, not just `DEAL_II_CONSTEXPR`. One benefit, for example, is creating `static constexpr Point<dim>` and `static constexpr std::array<Point<dim>, N>` without lots of macros and `#ifdef`s (such as all the ifdefs and code duplications that we used in `Physics::Elasticity::StandardTensors<dim>`).
1. A `constexpr` constructor does not allow a body containing `Assert`, thus `Assert` should be replaced with `static_assert`.

This PR wants to show what changes should be done to allow following these two related goals. As you noticed, these are possibly code-breaking changes in the user code especially in the case of `switch (dim)` and `(dim == 2) ? Point<dim>(x, y) : Point<dim>(x, y, z)`. Since you don't allow this incompatibility, I'll probably create `DEAL_II_CONSTEXPR` constructors and submit it as a separate small PR. Then I add some of the useful things here as separate PRs as well.

**Comment by bangerth** [comment]
I agree with the goal to make things `constexpr`, but I think for backward compatibility we need to preserve the ability to compile `Point<dim>(1,2)` even if `dim!=2` and the same for the 3-argument case. If we can't use `static_assert(dim==...)` and we can't use `Assert` in `constexpr` functions, what alternatives are there? Can we use `AssertThrow`?

Or, if we can't touch these functions, could we introduce another constructor that we can mark as `constexpr` that takes, for example, a `std::array<double,dim>`?

**Comment by masterleinad** [comment]
Looking at https://godbolt.org/z/LDSVVG, everything seems to work as expected with the C++14 relaxation for `constexpr` functions and I see the same behavior just using `constexpr` for the constructors with a recent compiler (`gcc-9.1.0`).
More precisely, I see
- compiling in a `constexpr` and a non-`constexpr` context works if the constrained is fulfilled,
- compiling in a non-`constexpr` context works if the constrained is not fulfilled (but will trigger an assertion at run time),
- compiling in a `constexpr` context fails if the constrained is not fulfilled with error message
```
dealii-new/include/deal.II/base/point.h: In function ‘int main()’:
dealii-new/tests/base/point_01.cc:47:27:   in ‘constexpr’ expansion of ‘dealii::Point<1>((double)1, (double)2)’
dealii-new/include/deal.II/base/point.h:394:3: error: call to non-‘constexpr’ function ‘void dealii::deal_II_exceptions::internals::issue_error_noreturn(dealii::deal_II_exceptions::internals::ExceptionHandling, const char*, int, const char*, const char*, const char*, ExceptionType) [with ExceptionType = dealii::StandardExceptions::ExcMessage]’
   Assert(dim == 2,
   ^~~~~~
```
Admittedly, this error message is more cryptic than using `static_constexpr`, but not impossible to understand.

In the end, we should be fine just using `DEAL_II_CONSTEXPR` (possibly checking `_cpp_constexpr >= 201304`, https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#cpp_constexpr_201304)

**Comment by rezarastak** [comment]
> Or, if we can't touch these functions, could we introduce another constructor that we can mark as constexpr that takes, for example, a std::array<double,dim>?

@bangerth I agree! interestingly, this idea came to me yesterday as well. We can build on the constructor `explicit Tensor(const array_type &)`.

Couple of related ideas:
1. May be helpful to replace the current definition of `Tensor::array_type` to use `std::array` instead of C style array.
2. May be helpful to remove the `explicit` specifier. It makes initializing `constexpr std::array<Tensor<rank, dim>, N>` or `constexpr std::array<Point<dim>, N>` much shorter and simpler.

**Comment by rezarastak** [comment]
@masterleinad Yes it is totally possible to define the constructors as `DEAL_II_CONSTEXPR` with existing `Assert` statements in the body. The downside is the verbosity for defining `static constexpr` variables. So, a good scenario would be to have at least one capable `constexpr` constructor, such as `constexpr Point<dim>(std::array<double, dim>)` as @bangerth mentioned. I use that constructor for `static constexpr` definitions. The rest of the constructors can be easily defined as `DEAL_II_CONSTEPXR` and be used in many other cases.

As an example, suppose I like to rewrite part of the `QTrapez` class as:
```c++
// quadrature_lib.h
template <>
class Qtrapez<1>{
  public:
  DEAL_II_CONSTEXPR const static std::array<Point<1>, 2> quadrature_points
#ifdef DEAL_II_HAVE_CXX14_CONSTEXPR_CAN_CALL_NON_CONSTEXPR
    {Point<1>{{0.}}, Point<1>{{1.}}}
#endif
    ;
  // The rest of the class
};

// quadrature_lib.cc
template <>
DEAL_II_CONSTEXPR const std::array<Point<1>, 2> Qtrapez<1>::quadrature_points
#ifndef DEAL_II_HAVE_CXX14_CONSTEXPR_CAN_CALL_NON_CONSTEXPR
    {Point<1>{{0.}}, Point<1>{{1.}}}
#endif
  ;
```
With a `DEAL_II_CONSTEXPR` constructor, that's the best (simplest and the most readable) I can write. On the other hand, if we have a `constexpr` constructor, we just write
```c++
// quadrature_lib.h
template <>
class Qtrapez<1>{
  public:
  constexpr static std::array<Point<1>, 2> quadrature_points
    {Point<1>{{0.}}, Point<1>{{1.}}};
    ;
  // The rest of the class
};

// quadrature_lib.cc
template <>
constexpr std::array<Point<1>, 2> Qtrapez<1>::quadrature_points;
```
and we enjoy its readability and the ease of maintaining it.

> possibly checking _cpp_constexpr >= 201304,

Do you think we should change the definition of the macros `DEAL_II_HAVE_CXX14_CONSTEXPR_CAN_CALL_NON_CONSTEXPR` and `DEAL_II_CONSTEXPR` based on this?

**Comment by rezarastak** [comment]
The discussion is very helpful here. I will submit a series of new PRs to follow up on it. I will close this PR once the discussion is done.

**Comment by masterleinad** [comment]
I don't have a strong opinion on the new constructor. On the one hand, I don't like having multiple functions that are doing the exact same thing because users never know which to prefer and why. In the end, the reasoning would be something like "use the new constructor if you can, otherwise use the old one". That doesn't really sound compelling to me if we were not to deprecate the old one (and I am not sure we want that right now). On the other hand, I see that this makes you write code more concise which is nice.
I think we agree on not forcing to use the new constructor everywhere because the old one does the job for non-`static` variables. Just create two pull requests and we can go from there.

I don't feel strongly about changing the feature check either. Currently, it does the job and I am not sure if the preprocessor macro is much better. I would only change the implementation if we discover a problem with the current one.

**Comment by bangerth** [comment]
I agree with everything @masterleinad says above. I'm just adamant about making sure that the existing code base continues to work -- which does not preclude us from adding constructors that allow us to be more concise and do things as `constexpr`.

As for changing the type of the `array_type` to `std::array`: This is probably pretty easy if you only consider `Tensor<1,dim>` and `Point<dim>`, but you'll have to consider how this type looks like for `Tensor<2,dim>` and higher.
